src/ControlSystem/FunctionVsTime.hpp

```diff
@@ -0,0 +1,172 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "ErrorHandling/Error.hpp"
+
+/// A function that has a piecewise-constant `MaxDeriv`th
+/// derivative.
+template <size_t MaxDeriv>
+class FunctionVsTimeNthDeriv {
+ public:
+  constexpr static size_t max_deriv() noexcept { return MaxDeriv; }
+
+  // this has outer length = number of components.
+  // each component has its own set of derivs.
+  using value_type = std::vector<std::array<double, MaxDeriv + 1>>;
+
+  FunctionVsTimeNthDeriv(double time,
+                         const value_type& initial_func_and_derivs);
+
+  FunctionVsTimeNthDeriv(const FunctionVsTimeNthDeriv&) = delete;
+  FunctionVsTimeNthDeriv& operator=(const FunctionVsTimeNthDeriv&) = delete;
+
+  /// Returns the function and its `MaxDeriv` derivatives at time t.
+  /// The function has multiple components.
+  value_type operator()(double t) const noexcept;
+  /// Updates the `MaxDeriv`th derivative of the function at the given time.
+  // updated_max_deriv is a vector of the `MaxDeriv`th to update
+  // for each component
+  void update(double time_of_update,
+              const std::vector<double>& updated_max_deriv) noexcept;
+
+ private:
+  // Return k such that mT[k] < t < mT[k+1].
+  // For t < mT[0], return 0.
+  // For t > mT[mT.size()-1] : return mT.size()-1
+  size_t ClosestTimeIndex(const double t) const;
+
+  // history of the function times and derivs
+  std::vector<double> times_;
+  std::vector<value_type> derivs_;
+};
+
+/// Function implementations
+// constructor, initialize times and derivs
+template <size_t MaxDeriv>
+FunctionVsTimeNthDeriv<MaxDeriv>::FunctionVsTimeNthDeriv(
+    double time, const value_type& initial_func_and_derivs)
+    : times_{time}, derivs_{initial_func_and_derivs} {}
+
+// operator, returns function and derivs at time t
+template <size_t MaxDeriv>
+typename FunctionVsTimeNthDeriv<MaxDeriv>::value_type
+FunctionVsTimeNthDeriv<MaxDeriv>::operator()(double t) const noexcept {
+
+  const size_t k = ClosestTimeIndex(t);
+  const double dt = t - times_[k];
+
+  // initialize the return variable
+  // (with derivs_, since gets overwritten anyways)
+  value_type result = derivs_[k];
+
+  // loop over components
+  // (the number of components is same size as what is at the end of derivs)
+  for (size_t i = 0; i < derivs_[k].size(); i++) {
+    // loop over derivs
+    for (size_t d = 0; d < MaxDeriv; ++d) {
+      result[i][d] = derivs_[k][i][MaxDeriv];
+      for (size_t dd = MaxDeriv; dd --> d ;) {
+        result[i][d] *= dt / (dd - d + 1);
```

Comment:
benchmark compare this to storing an std::array<double,MaxDeriv> factors indexed by [dd-d]